enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,610 @@
+                        items:
+                          description: GCPResourceLabel is a label to apply to GCP resources created for the cluster.
+                          properties:
+                            key:
+                              description: key is the key part of the label. A label key can have a maximum of 63 characters and cannot be empty. Label must begin with a lowercase letter, and must contain only lowercase letters, numeric characters, and the following special characters `_ -`.
+                              maxLength: 63
+                              minLength: 1
+                              pattern: ^[a-z][0-9a-z_-]+$
+                              type: string
+                            value:
+                              description: value is the value part of the label. A label value can have a maximum of 63 characters and cannot be empty. Value must contain only lowercase letters, numeric characters, and the following special characters `_ -`.
+                              maxLength: 63
+                              minLength: 1
+                              pattern: ^[0-9a-z_-]+$
+                              type: string
+                          required:
+                            - key
+                            - value
+                          type: object
+                        maxItems: 32
+                        type: array
+                        x-kubernetes-validations:
+                          - message: resourceLabels are immutable and may only be configured during installation
+                            rule: self.all(x, x in oldSelf) && oldSelf.all(x, x in self)
+                      resourceTags:
+                        description: resourceTags is a list of additional tags to apply to GCP resources created for the cluster. See https://cloud.google.com/resource-manager/docs/tags/tags-overview for information on tagging GCP resources. GCP supports a maximum of 50 tags per resource.
+                        items:
+                          description: GCPResourceTag is a tag to apply to GCP resources created for the cluster.
+                          properties:
+                            key:
+                              description: key is the key part of the tag. A tag key can have a maximum of 63 characters and cannot be empty. Tag key must begin and end with an alphanumeric character, and must contain only uppercase, lowercase alphanumeric characters, and the following special characters `. _ -`.
+                              maxLength: 63
+                              minLength: 1
+                              pattern: ^[a-zA-Z0-9]([0-9A-Za-z_.-]{0,61}[a-zA-Z0-9])?$
+                              type: string
+                            value:
```

Comment:
I believe, technically, GCP tags have a 1:many relationship between key:value, but this install config model has a 1:1 relationship. Could this potentially be problematic? Could users want to supply more than one value per key?

Reply:
We can create a maximum of 1000 keys and for each key 1000 values can be created. But OpenShift will not create tags, and only add tags to the resources it creates, I think tags with key and a value should suffice.

Reply:
When creating a tag key, multiple tag values can be defined for a key and user can choose one of these predefined values only for a key when attaching to a resource. User cannot assign any random key or value to a resource, but has to choose from one of the predefined keys and the predefined values for that key.

Below example lists the values defined for the key `ocp_tag_dev` created in Organization having ID `54643501348`
```
$ gcloud resource-manager tags values list --parent=54643501348/ocp_tag_dev
NAME                       SHORT_NAME  DESCRIPTION
tagValues/281478803580207  bar
tagValues/281483034419673  foo
tagValues/281484411361997  baz
```
Add a user to add a tag with key `ocp_tag_dev` then can choose one of `foo, bar, baz`as the value for the tag.
```
$ gcloud resource-manager tags bindings create --parent=//compute.googleapis.com/projects/openshift-gce-devel/zones/asia-south1-a/instances/7120657451740199772 --tag-value=54643501348/ocp_tag_dev/foo --location=asia-south1-a
```
And to add the same through OpenShift, will have to provide list of key:value pairs in the install-config.

Reply:
We can tag a resource with only one value for a key, which is documented [here](https://cloud.google.com/resource-manager/docs/tags/tags-creating-and-managing#attaching). We can tag a resource with the key:value pair `ocp_tag_dev:foo` or `ocp_tag_dev:bar` or `ocp_tag_dev:baz`. The install-config would look like below to tag resources with value `foo`
```
platform:
  gcp:
    organizationID: 54643501348
    userTags:
      ocp_tag_dev: foo
```

Reply:
Ok, making more sense now thank you

I think the API should actually be as below to be consistent with kube standards right?
```
platform:
  gcp:
    organizationID: 54643501348
    userTags:
     - name: ocp_tag_dev
       value: foo
```

Reply:
Yeah, tags in Infrastructure will be an array of struct type with `key` and `value` as members, in install-config it will be type `map[string]string`.

Reply:
Why not make them the same?

Reply:
I have observed for key:value parameters in install-config go maps are used and same is the type for AWS, Azure tags as well and decided to keep the same to be uniform across solutions. Received a [comment](https://github.com/openshift/enhancements/pull/1217#discussion_r982979140) as well to keep it map type.

Reply:
comment incorporated.